### PR TITLE
Fix path issues with self-contained documents

### LINKED ISSUE #286: Path arguments do not have the expected values.
Path arguments are lazily evaluated *after* the working directory is changed, so any use of e.g. `normalizePath` resolves relative to the *new* working directory, not the *old* working directory.

Example:

```r
# This creates an output at path/to/source/path/to/even/deeper/output/tst.html
render("path/to/source/tst.Rmd",
  intermediates_dir = {normalizePath("path/intermediate/")},
  output_dir = {normalizePath("path/to/even/deeper/output/")},
  clean=TRUE)

# Whereas this creates it at path/to/even/deeper/output/tst.html
intermediates_dir = {normalizePath("path/intermediate/")};
output_dir = {normalizePath("path/to/even/deeper/output/")}
render("path/to/source/tst.Rmd",
  intermediates_dir = intermediates_dir,
  output_dir = output_dir,clean=TRUE)
```

### LINKED ISSUE #429: images not found when rendering self contained document using intermediates dir
Minimal RMarkdown document using the logo from rstudio.com: //www.rstudio.com/wp-content/uploads/2014/03/blue-250.png
```
![rstudio](blue-250.png)
```

Rendering command including console output:
```
> rmarkdown::render("self_contained.Rmd",intermediates_dir="/tmp",output_options=list(self_contained=TRUE))


processing file: self_contained.Rmd
  |.................................................................| 100%
  ordinary text without R code


output file: /tmp/self_contained.knit.md

/Applications/RStudio.app/Contents/MacOS/pandoc/pandoc /tmp/self_contained.utf8.md --to html --from markdown+autolink_bare_uris+ascii_identifiers+tex_math_single_backslash-implicit_figures --output self_contained.html --smart --email-obfuscation none --self-contained --standalone --section-divs --template /Library/Frameworks/R.framework/Versions/3.2/Resources/library/rmarkdown/rmd/h/default.html --variable 'theme:bootstrap' --include-in-header /var/folders/ts/s940qvdj5vj1czr9qh07fvtw0000gn/T//Rtmp8lqgQv/rmarkdown-str12e934bd9affc.html --mathjax --variable 'mathjax-url:https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --no-highlight --variable highlightjs=/Library/Frameworks/R.framework/Versions/3.2/Resources/library/rmarkdown/rmd/h/highlight 
pandoc: Could not fetch blue-250.png
blue-250.png: openBinaryFile: does not exist (No such file or directory)
Error: pandoc document conversion failed with error 67
```

Output from 

`getwd()` is the directory containing both the RMarkdown document and the logo image.

### PR DESCRIPTION
**Pull request by jmcphers**
This change fixes #429 by copying a document's supporting resources to the intermediates folder prior to invoking Pandoc, when the following are both true:

- The document is self-contained.
- The intermediates folder has been explicitly specified.

It also fixes #286, and by extension #130, by resolving `intermediates_dir` and `output_dir` before changing the working directory. 

There are two possible breaking changes here:

- `output_format` now has a new argument (`intermediates_generator`), which I've placed in the arg list between pre- and post-processors. I think this is the most logical place for it since that's when it runs, but if there's any existing third-party usage of `output_format` that uses positional args and supplies pre- and post-processors, it will break. If we think this might be the case I'll move the new arg to the end.

- `intermediates_dir` and `output_dir` are now resolved against the current working directory when calling `::render`, not the directory in which the input file resides. Obviously several people think the old behavior was a bug (and it is counter-intuitive) but it's possible that third-party code somewhere relies on it.

### CODE REVIEW

**Comment by jjallaire** [comment]
I'm not at all worried about the positional argument issue (I think everyone uses these named). I also think the new behavior is (a) More intuitive; and (b) In any case previously unspecified so it's not really breaking a contract. Merging as-is.
